sycl/plugins/unified_runtime/ur/adapters/level_zero/usm.cpp

```diff
@@ -673,42 +630,91 @@ enum umf_result_t USMMemoryProvider::free(void *Ptr, size_t Size) {
   return UMF_RESULT_SUCCESS;
 }
 
-void USMMemoryProvider::get_last_native_error(const char **ErrMsg,
-                                              int32_t *ErrCode) {
-  (void)ErrMsg;
-  *ErrCode = static_cast<int32_t>(getLastStatusRef());
+umf_result_t L0MemoryProvider::GetL0MinPageSize(void *Mem, size_t *PageSize) {
+  ur_result_t Res = UR_RESULT_SUCCESS;
+  void *Ptr = Mem;
+
+  if (!Mem) {
+    Res = allocateImpl(&Ptr, 1, 0);
+    if (Res != UR_RESULT_SUCCESS) {
+      goto err_set_status;
+    }
+  }
+
+  // Query L0 for the minimal page size.
+  Res = USMQueryPageSize(Context, Ptr, PageSize);
+  if (Res != UR_RESULT_SUCCESS) {
+    goto err_dealloc;
+  }
+
+  if (!Mem) {
+    Res = USMFreeImpl(Context, Ptr);
+    if (Res != UR_RESULT_SUCCESS) {
+      goto err_set_status;
+    }
+  }
+
+  return UMF_RESULT_SUCCESS;
+
+err_dealloc:
+  if (!Mem) {
+    USMFreeImpl(Context, Ptr);
+  }
+err_set_status:
+  getLastStatusRef() = Res;
+  return UMF_RESULT_ERROR_MEMORY_PROVIDER_SPECIFIC;
 }
 
-umf_result_t USMMemoryProvider::get_min_page_size(void *Ptr, size_t *PageSize) {
-  (void)Ptr;
+umf_result_t L0MemoryProvider::get_min_page_size(void *Ptr, size_t *PageSize) {
+  std::ignore = Ptr;
+
+  // Query L0 for min page size. Use provided 'Ptr'.
+  if (Ptr) {
+    return GetL0MinPageSize(Ptr, PageSize);
```

Comment:
why did we move this from initialize to query method? Problem is that GetL0MinPageSize is not caching the value, so every time the user makes a query for min_page_size, we could have an unexpected allocation, right?

Reply:
oh yes, I see the caching. thanks.